homeassistant/components/aemet/weather_update_coordinator.py

```diff
@@ -0,0 +1,652 @@
+        return None
+
+
+class TownNotFound(UpdateFailed):
+    """Raised when town is not found."""
+
+
+class WeatherUpdateCoordinator(DataUpdateCoordinator):
+    """Weather data update coordinator."""
+
+    def __init__(self, aemet, latitude, longitude, forecast_mode, hass):
+        """Initialize coordinator."""
+        super().__init__(
+            hass, _LOGGER, name=DOMAIN, update_interval=WEATHER_UPDATE_INTERVAL
+        )
+
+        self._aemet = aemet
+        self._station = None
+        self._town = None
+        self._latitude = latitude
+        self._longitude = longitude
+        self._forecast_mode = forecast_mode
+        self._data = {
+            "daily": None,
+            "hourly": None,
+            "station": None,
+        }
+
+    async def _async_update_data(self):
+        data = {}
+        with async_timeout.timeout(120):
+            weather_response = await self._get_aemet_weather()
+            data = self._convert_weather_response(weather_response)
+        return data
+
+    async def _get_aemet_weather(self):
+        """Poll weather data from AEMET OpenData."""
+        weather = await self.hass.async_add_executor_job(self._get_weather_and_forecast)
+        return weather
+
+    def _get_weather_station(self):
+        if not self._station:
+            self._station = (
+                self._aemet.get_conventional_observation_station_by_coordinates(
+                    self._latitude, self._longitude
+                )
+            )
+            if self._station:
+                _LOGGER.debug(
+                    "station found for coordinates [%s, %s]: %s",
+                    self._latitude,
+                    self._longitude,
+                    self._station,
+                )
+        if not self._station:
+            _LOGGER.debug(
+                "station not found for coordinates [%s, %s]",
+                self._latitude,
+                self._longitude,
+            )
+        return self._station
+
+    def _get_weather_town(self):
+        if not self._town:
+            self._town = self._aemet.get_town_by_coordinates(
+                self._latitude, self._longitude
+            )
+            if self._town:
+                _LOGGER.debug(
+                    "town found for coordinates [%s, %s]: %s",
+                    self._latitude,
+                    self._longitude,
+                    self._town,
+                )
+        if not self._town:
+            _LOGGER.error(
+                "town not found for coordinates [%s, %s]",
+                self._latitude,
+                self._longitude,
+            )
+            raise TownNotFound
+        return self._town
+
+    def _get_weather_and_forecast(self):
+        """Get weather and forecast data from AEMET OpenData."""
+
+        self._get_weather_town()
+
+        daily = None
+        if self._forecast_mode == FORECAST_MODE_DAILY:
+            daily = self._aemet.get_specific_forecast_town_daily(
```

Comment:
This class is very convoluted because you have three different api calls.  Can these be separated into separate UpdateCoordinator's?  This would be ideal in the case that you have multiple entity types that require data from different api calls, which seems to be the case.  Otherwise, a failure in one api call results in unavailable data for everything.

Reply:
I need to investigate this.

Reply:
@MatthewFlamm I'm a bit of a newbie here, so I'll need some help to achieve this...
Can you provide examples of other integrations which have multiple update coordinators?
With the other changes requested, AbstractAemetSensor and AemetWeather inherit from CoordinatorEntity, thus containing a single coordinator, so how can I add multiple coordinators now?

Reply:
@MartinHjelmare @MatthewFlamm I'm a bit of a newbie here, so I'll need some help to achieve this...
Can you provide examples of other integrations which have multiple update coordinators?
With the other changes requested, AbstractAemetSensor and AemetWeather inherit from CoordinatorEntity, thus containing a single coordinator, so how can I add multiple coordinators now?

Reply:
NWS is one example:
https://github.com/home-assistant/core/blob/a67b598971eb5142751dd6642697257b82aa31aa/homeassistant/components/nws/__init__.py#L112-L146

Reply:
Each entity will only be connected to one coordinator.

Reply:
Although NWS actually has two coordinators per entity.

Reply:
> Although NWS actually has two coordinators per entity.

Am I allowed to implement multiple coordinators per entity then?

Reply:
I think the suggestion here is to make one coordinator per forecast mode and this will map 1:1 to the different weather entities which each has a different forecast mode.

Reply:
> I think the suggestion here is to make one coordinator per forecast mode and this will map 1:1 to the different weather entities which each has a different forecast mode.

@MartinHjelmare indeed, that would work for forecast mode, but there are some problems with that:
- Station data replaces predicted hourly data, so we would need on coordinator for hourly forecast and another one for station.
- Sensor/Weather entities created for daily mode won't have precise values, since daily data si split into 6h intervals:
```
"temperatura" : {
        "maxima" : 2,
        "minima" : -1,
        "dato" : [ {
          "value" : -1,
          "hora" : 6
        }, {
          "value" : 0,
          "hora" : 12
        }, {
          "value" : 1,
          "hora" : 18
        }, {
          "value" : 1,
          "hora" : 24
        } ]
      },
```
- So, in the end, for the current temperature, humidity, etc. values on daily weather, we need to rely on the hourly coordinator.

Therefore, the daily mode would need 3 coordinators: `daily, station and hourly`, and the hourly mode would need 2 coordinators: `hourly and station`.

Reply:
Yeah, I haven't looked into details if this would make things more readable or not. My suggestion is that we save this work for a potential follow up PR.

Reply:
> Yeah, I haven't looked into details if this would make things more readable or not. My suggestion is that we save this work for a potential follow up PR.

Ok, so I'll leave it for a potential follow up PR and remove the config_mode creating two weather entities with hourly and daily forecast.

Reply:
Yeah, remove the config flow option and have one entity enabled by default and the other entity disabled by default.

Reply:
@MartinHjelmare done in https://github.com/home-assistant/core/pull/45074/commits/53af241cc5b9b553758b0afdc30646feca831f4c

Reply:
NWS has a similar, although it seems not as convoluted scheme for the data.  There is 1 coordinator for the current observations, and 1 coordinator for each forecast type.  Each entity gets passed 2 coordinators, the current observation and the relevant forecast coordinator.  This only makes sense if you can use separate the logic into separate coordinators.  If not, then this is a good reason to reject this structure.

I agree that given the complexity this should be a follow-on PR, if possible, to make the integration even better.